### ISSUE TITLE
SIP Contact Header on recieving calls from providers

### ISSUE THREAD

**Issue by gil-obradors** [irontec/ivozprovider#1285]
### Description

I have observed that the SIP/180 or SIP/183 reply to an INVITE from call-provider is filled with :

_(incoming call from one provider, that belongs to retail account)_
```Contact: <sip:xxxxxxxx@PRIVATE LAN RETAIL ACCOUNT:5060;alias=1PULIC IP RETAIL ACCOUNT3.8~5060~1>```

_(in case of outgoing call from retail acount to provider)_
``` <sip:asterisk@127.0.0.1:6060> ```

_(in case of residencial incoing call from one provider)_
```Contact: <sip:127.0.0.1:6060>```

- Must our peers know anything about internal info?
- Is this real contact between a sip proxy?
- the contact must be the proxy or the final UA?

I trying to read alot to understand SIP call flows and RFC but I haven't got experience about CONTACT field, and proxys contributions to this,  only I can contribute with:
- Two Spanish providers showing in this field, public ip and port from proxy. And their don't pay attention to this field ( because the call flow going well, calls OK)
- One Spanish provider refusing replys from INVITES (incoming call for me, outgoing call for them) because this header isn't correctly for them.

Here I'm to contribute with more info if needed
Thanks a lot.

---

- IvozProvider version: 
``` 2.11~2.11.1~20190711.2151~8cf11ef+artemis all ```
- Related sections: 
    - [x] SIP & Telephony
    - [ ] vPBX Functionalities
    - [ ] Billing & Invoicing
    - [ ] Web Interface
    - [ ] API
    - [ ] Other

### Troubleshooting

Callflow crash ( says my provider)

**Comment by cruzccl** [comment]
Hi @gil-obradors,

Unfortunately SIP routing is not simple and RFC3261 must be read thoroughly  to understand all use cases.

In IvozProvider, vpbx/residential calls traverse Asterisk, that acts as B2BUA and talks to carriers though a SIP Proxy. SIP proxy inserts itself in dialog's Route Set (thanks to Record-route header of initial transactions), making Contact inserted by Asterisk irrelevant for carriers. In these scenarios (**non-first hop** in carrier's perspective) future within dialog transaction must be routed to address in Record-Route instead (inserting Contact URI in R-URI).

Why do some carriers analyze Contact header's address? Because when they receive INVITEs as **first hop** (without any proxy in the middle), address in Contact header will be where within dialogs future request should go. In these scenarios (a majority for them), Contact address must be reachable. That's why they drop calls with non reachable addresses in Contact header.

To sum up, in IvozProvider, carriers should not analyze Contact address as a proxy is in the middle of the signalling path and they will never talk that address (they will talk to address in Record-Route instead).

Regards,

**Comment by gil-obradors** [comment]
Thanks @cruzccl  for your rich answer and time.
